### PR TITLE
LCD-14301 Make serviceAccountKey optional

### PR DESCRIPTION
**Pull request by briangreenwald**
### Background

https://issues.liferay.com/browse/LCD-14380

When no credentials are explicitly set, Google's libraries automatically attempt to federate access control as follows:

- If the environment variable GOOGLE_APPLICATION_CREDENTIALS is set, ADC uses the service account key or configuration file that the variable points to.
- If the environment variable GOOGLE_APPLICATION_CREDENTIALS isn't set, ADC uses the service account that is attached to the resource that is running your code. (workload identity)

As a best practice, Liferay Cloud is moving to leverage Workload Identity for access control across our services that interact with GCP directly. As we are now looking to move our Document Library to GCSStore for all customers, our Liferay service will now be doing just that.

This change can thus also be used by anyone running Liferay in a GKE and is not strictly limited to our use-case in Cloud.

### CODE REVIEW

**Comment by briangreenwald** [comment]
@brandizzi Can you please share your test steps here? Thanks!

**Comment by catarineaguiar** [comment]
@briangreenwald no regressions found.

**Comment by brandizzi** [comment]
This pull request refers to [LPS-165561](https://issues.liferay.com/browse/LPS-165561).

## How to test GCS Store with Application Default Credentials

This test relies on ADCs alone. It is especially useful to work on the Liferay changes themselves, and is probably much easier to follow by colleagues outside Cloud.

### Prepare Google Cloud resources
Supposing you already have a Google Cloud account:
1. Create a bucket ([reference](https://cloud.google.com/storage/docs/creating-buckets)];
    * Take note of the bucket name. In our examples, we are going to call it `abrandizzi-lcd14301-20220916`; replace it with the name you choose.
2. Set up [Application Default Credentials](https://cloud.google.com/docs/authentication/provide-credentials-adc#local-dev) using the same account you used to create the bucket.

### Prepare Liferay
Given a working, compatible, Liferay portal bundle (ideally built from `ant all` from the same branch with these changes):
1. Remove the contents from `osgi/state`;
2. Deploy the altered GCSStore bundle into a portal bundle, inside `modules/portal`;
   * It is fundamental to copy the JAR file directly into `modules/portal`. Otherwise, an older version of the bundle can start up, create some configuration and pollute the `osgi/state` directory. In our experience, OSGi configs fail to load the proper configuration when they are hot-deployed.
3. Create the file `osgi/configs/com.liferay.portal.store.gcs.GCSStoreConfiguration.config` with the following content:
    ```
    bucketName="abrandizzi-lcd14301-20220916"  # Replace with the real bucket name
    aes256Key=""  # This value is irrelevant for our tests, but GCS Store requires it to be set.
    ```
### Check if the configuration works

1. Start up Liferay;
2. Observe if there is no error message on startup;
3. Log into Liferay;
4. Go to `Control Panel > System Settings > File Storage > GCS Store`. Confirm that
    * the bucket name you've given is set there; and
    * there is no service account key:
    gcs-store-config
5. Go to `Control Panel > Server Adminstration  > Properties > Portal Properties`; search for `dl.store.imp` Confirm that the value is `com.liferay.portal.store.gcs.GCSStore`.
    gcs-store-enabled

### Use Documents & Media and get the necessary data

1. Go to the `Sidebar > Content > Documents and Media`;
2. Create a new file by upload
    * Here we uploaded the file named `administrator` in this screenshot:
        Captura de tela de 2022-09-20 18-26-00
3. Get the `fileEntryId` from the URL of the new file.
4. In the `lportal` database, search for the `repositoryId` of that `fileEntryId`
    ```
    $ mysql> select companyId, fileEntryId, folderId, repositoryId, name, fileName, version from DLFileEntry where fileEntryId = 44935 order by fileEntryId;
    +-----------+-------------+----------+--------------+------+--------------------+---------+
    | companyId | fileEntryId | folderId | repositoryId | name | fileName           | version |
    +-----------+-------------+----------+--------------+------+--------------------+---------+
    |     20098 |       44935 |        0 |        20122 | 203  | administgrator.jpg | 1.0     |
    +-----------+-------------+----------+--------------+------+--------------------+---------+
    ```

### Check the bucket
1. Go to the bucket you have created
2. The first folder should be the `companyId`; click on it and then on the folder with the same `repositoryId` you've found above.
3. Click in the folder with the same `name` found above (in our case, `203`).
4. Download the file named as the version above (in our case, `1.0`).
    dl-file-in-bucket
5. Confirm it is the same file you have uploated.

**Comment by brandizzi** [comment]
@tinatian please let us know of anything that should be adjusted. Thanks!

File: modules/apps/portal-store/portal-store-gcs/src/main/java/com/liferay/portal/store/gcs/GCSStore.java [hunk 1]
```diff
@@ -315,6 +315,27 @@ private String _getFileVersionKey(
...
+				_log.info(
+					"No credentials set for GCS Store. Library will default " +
+						"to using Application Default Credentials or " +
+							"Workload Identity for auth");
+			}
+
+			return ServiceAccountCredentials.getApplicationDefault();
+		}
+
+		try (InputStream inputStream = new ByteArrayInputStream(
```
Comment (dantewang) [inline modules/apps/portal-store/portal-store-gcs/src/main/java/com/liferay/portal/store/gcs/GCSStore.java:331]: No need to try-with-resource on `ByteArrayInputStream`. Its `close()` method is actually empty. Please inline this variable.

File: modules/apps/portal-store/portal-store-gcs/src/main/java/com/liferay/portal/store/gcs/GCSStore.java [hunk 2]
```diff
@@ -386,59 +407,37 @@ private void _initEncryption() {
...
-					_gcsStoreConfiguration.retryDelayMultiplier()
-				).setRpcTimeoutMultiplier(
-					_gcsStoreConfiguration.rpcTimeoutMultiplier()
-				).build()
-			);
-
-		if (_googleCredentials != null) {
-			storageOptionsBuilder.setCredentials(_googleCredentials);
+		try {
+			_googleCredentials = _getGoogleCredentials();
```
Comment (dantewang) [inline modules/apps/portal-store/portal-store-gcs/src/main/java/com/liferay/portal/store/gcs/GCSStore.java:410]: No need to extract a separate method here. Please inline.

**Review by dantewang — CHANGES_REQUESTED** [review]
@briangreenwald The first 2 commits modify several things, but half of the things modified in the first commit is restored in the second commit. Usually we don't want such history to be kept.

Please update the PR as

1. An SF commit to remove try-with-resource on `ByteArrayInputStream`
2. 1 or 2 commit to make service account key optional and fallback to the application default. If using 2 commits,
  a. Make the configuration optional in the first commit
  b. Apply by adding fallback in the second commit
3. The commit to add test
4. Merge the last commit that modifies the log message into 2 or 2.b

File: modules/apps/portal-store/portal-store-gcs/src/main/java/com/liferay/portal/store/gcs/GCSStore.java (hunk 2 above)
Comment (brandizzi) [reply modules/apps/portal-store/portal-store-gcs/src/main/java/com/liferay/portal/store/gcs/GCSStore.java:410]: Thank you for your feedback, @dantewang! The PR has be re-sent as https://github.com/liferay-core-infra/liferay-portal/pull/2153.

**Comment by briangreenwald** [comment]
See #2153

**Pull request closed without merging**